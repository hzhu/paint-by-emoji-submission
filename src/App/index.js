import React, { useState, useRef, useEffect } from "react";
import EmojiPicker from "../emoji-picker";
import {
  MODE,
  EMOJIS,
  MIN_WIDTH,
  MIN_HEIGHT,
  MAX_WIDTH,
  MAX_HEIGHT,
  DEFAULT_WIDTH,
  DEFAULT_HEIGHT
} from "../constants";
import { copyToClipboard } from "../util";

import "./index.css";

const App = () => {
  const [mode, setMode] = useState();
  const [width, setWidth] = useState(DEFAULT_WIDTH);
  const [height, setHeight] = useState(DEFAULT_HEIGHT);
  const [isPainting, setIsPainting] = useState(false);
  const [activeEmoji, setActiveEmoji] = useState("😀");
  const [grid, setGrid] = useState(
    [...Array(DEFAULT_HEIGHT)].map(() =>
      [...Array(DEFAULT_WIDTH)].map(() => "")
    )
  );
  const [emptyGrid, setEmptyGrid] = useState(false);

  const onHeightChange = e => {
    if (!e.target.value) return;
    const newGrid = JSON.parse(JSON.stringify(grid));
    const currHeight = grid.length;
    const nextHeight = Number(e.target.value);
    setHeight(nextHeight);
    if (nextHeight > MAX_HEIGHT || nextHeight < MIN_HEIGHT) return;
    if (nextHeight > currHeight) {
      const rowsToAdd = nextHeight - currHeight;
      for (let i = 0; i < rowsToAdd; i++) {
        const width = newGrid[0].length;
        newGrid.push([...Array(width)].map(() => ""));
      }
    } else {
      const rowsToRemove = currHeight - nextHeight;
      for (let i = 0; i < rowsToRemove; i++) {
        newGrid.pop();
      }
    }
    setGrid(newGrid);
  };

  const onWidthChange = e => {
    if (!e.target.value) return;
    const newGrid = JSON.parse(JSON.stringify(grid));
    const currWidth = grid[0].length;
    const nextWidth = Number(e.target.value);
    setWidth(nextWidth);
    if (nextWidth > MAX_WIDTH || nextWidth < MIN_WIDTH) return;
    if (nextWidth > currWidth) {
      const colsToAdd = nextWidth - currWidth;
      for (let i = 0; i < colsToAdd; i++) {
        newGrid.forEach(row => row.push(""));
      }
    } else {
      const colsToRemove = currWidth - nextWidth;
      for (let i = 0; i < colsToRemove; i++) {
        newGrid.forEach(row => row.pop());
      }
    }
    setGrid(newGrid);
  };

  return (
    <div className="app">
      <div className="emoji-paint tc">
        <EmojiToolbar
          mode={mode}
          width={width}
          height={height}
          setMode={setMode}
          activeEmoji={activeEmoji}
          setActiveEmoji={setActiveEmoji}
          onWidthChange={onWidthChange}
          onHeightChange={onHeightChange}
        />
        <ModePanel mode={mode} />
        <EmojiGrid
          grid={grid}
          mode={mode}
          emptyGrid={emptyGrid}
          setEmptyGrid={setEmptyGrid}
          activeEmoji={activeEmoji}
          isPainting={isPainting}
          setIsPainting={setIsPainting}
        />
        <div className="flex justify-end mh3">
          <button
            onClick={() => setEmptyGrid(true)}
            className="ph3 br2 fw5 pointer"
            style={{
              border: "1px solid #bfc0c0"
            }}
          >
            Clear
          </button>
          <button
            onClick={() => copyToClipboard(grid)}
            className="f6 fw6 ml2 white ph3 pv3 bn br2 pointer"
            style={{ backgroundColor: "rgb(52, 175, 127)" }}
          >
            Copy to clipboard
          </button>
        </div>
        <div className="pa3">
          <span>
            Learn how to add a <em>:blank:</em> emoji to your Slack workspace{" "}
            <a
              className="link bg-animate hover-bg-lightest-blue"
              href="https://get.slack.help/hc/en-us/articles/206870177-Add-custom-emoji"
            >
              here
            </a>
            !
          </span>
        </div>
      </div>
    </div>
  );
};

const EmojiToolbar = ({
  mode,
  width,
  height,
  setMode,
  activeEmoji,
  setActiveEmoji,
  onWidthChange,
  onHeightChange
}) => (
  <div className="emoji-paint__toolbar">
    <div className="emoji-paint__controls">
      <EmojiPicker
        value={activeEmoji}
        onSelect={emoji => setActiveEmoji(emoji)}
      />
      <button
        onClick={() => setMode(MODE.brush)}
        className={`emoji-paint__control ${
          mode === MODE.brush ? "b--green bw1" : ""
        }`}
      >
        <img
          className="emoji-paint__control_icon"
          src="brush.png"
          alt="brush"
        />
      </button>
      <button
        onClick={() => setMode(MODE.erase)}
        className={`emoji-paint__control ${
          mode === MODE.erase ? "b--green bw1" : ""
        }`}
      >
        <img
          className="emoji-paint__control_icon"
          src="eraser.png"
          alt="eraser"
        />
      </button>
    </div>
    <div>
      <label className="emoji-paint__dimension">
        Width
        <input
          min={MIN_WIDTH}
          max={MAX_WIDTH}
          type="number"
          className="emoji-paint__dimension_input"
          onChange={onWidthChange}
          defaultValue={DEFAULT_WIDTH}
        />
      </label>
      <label className="emoji-paint__dimension">
        Height
        <input
          min={MIN_HEIGHT}
          max={MAX_HEIGHT}
          type="number"
          className="emoji-paint__dimension_input"
          onChange={onHeightChange}
          defaultValue={DEFAULT_HEIGHT}
        />
      </label>
      <div className="dark-red pt2 tl">
        <div aria-live="polite">
          {width > MAX_WIDTH ? (
            <span data-testid="bounds-error">
              <EmojiIcon emoji="⚠️" /> The maximum width is {MAX_WIDTH}
            </span>
          ) : width < MIN_WIDTH ? (
            <span data-testid="bounds-error">
              <EmojiIcon emoji="⚠️" /> The minimum width is {MIN_WIDTH}
            </span>
          ) : null}
        </div>
        <div aria-live="polite">
          {height > MAX_HEIGHT ? (
            <span data-testid="bounds-error">
              <EmojiIcon emoji="⚠️" /> The maximum height is {MAX_HEIGHT}
            </span>
          ) : height < MIN_HEIGHT ? (
            <span data-testid="bounds-error">
              <EmojiIcon emoji="⚠️" /> The minimum height is {MIN_HEIGHT}
            </span>
          ) : null}
        </div>
      </div>
    </div>
  </div>
);

const EmojiIcon = ({ emoji }) => (
  <span role="img" aria-label={`${EMOJIS[emoji].name} emoji`}>
    {emoji}
  </span>
);

const Comp = React.forwardRef(({ m, n, emptyGrid, setEmptyGrid }, ref) => {
  const [content, setContent] = useState("");

  useEffect(() => {
    if (emptyGrid) {
      setContent("");
      setEmptyGrid(false);
    }
  }, [emptyGrid, setEmptyGrid]);

  const onMouseDown = () => {
    if (ref.current.mode === MODE.brush) {
      setContent(ref.current.activeEmoji);
    } else {
      setContent("");
    }
  };

  const onMouseEnter = () => {
    if (ref.current.pressed) {
      if (ref.current.mode === MODE.brush) {
        setContent(ref.current.activeEmoji);
      } else {
        setContent("");
      }
    }
  };
  return (
    <span
      key={n}
      role="img"
      className="pa1"
      aria-label={EMOJIS[content].name}
      data-testid={`cell-${m}-${n}`}
      style={{
        width: "45px",
        height: "45px",
        fontSize: "37px",
        userSelect: "none",
        borderRight: "1px solid #d5d5d6",
        borderBottom: "1px solid #d5d5d6"
      }}
      onMouseDown={onMouseDown}
      onMouseEnter={onMouseEnter}
    >
      {content}
    </span>
  );
});
function areEqual(prevProps, nextProps) {
  // See: https://reactjs.org/docs/react-api.html#reactmemo

  // React.memo() is used to memoize a component by shallow comparing it's props.
  // Here, it is used as an escape hatch so that each cell only renders based
  // on its internal component state. State updates to parent components
  // will not cause a render to the cells if `true` is returned from this `areEqual`
  // function.

  // This is improves rendering performance if the canvas is a 500x500 grid.
  // We do not want to re-render 250,000 cells upon parent component state
  // changes. For example, if the user selects the brush mode, and state for
  // <EmojiPaint /> updates, it forces the entire subtree to re-render. That
  // subtree could contain 250,000 grid cell components.

  if (prevProps.emptyGrid !== nextProps.emptyGrid) {
    // We want to control when cells reset their state when a user clicks the
    // "Clear" button. Returning `false` here forces the children to render
    // to it's initial empty state.
    return false;
  }

  return true;
}
const GridCell = React.memo(Comp, areEqual);

const EmojiGrid = ({ grid, mode, activeEmoji, emptyGrid, setEmptyGrid }) => {
  const ref = useRef({});

  useEffect(() => {
    ref.current.mode = mode;
    ref.current.activeEmoji = activeEmoji;
  }, [mode, activeEmoji]);

  return (
    <div
      data-testid="grid"
      className={`ma3 br3 dib ${mode ? "pointer" : "not-allowed"}`}
      style={{
        border: "1px solid #a0a0a2"
      }}
      onMouseUp={() => (ref.current.pressed = false)}
      onMouseDown={() => (ref.current.pressed = true)}
      onMouseLeave={() => (ref.current.pressed = false)}
      onContextMenu={() => (ref.current.pressed = false)}
    >
      {grid.map((row, m) => (
        <div key={m} data-testid={`row-${m}`} className="flex">
          {row.map((cell, n) => (
            <GridCell
              m={m}
              n={n}
              key={n}
              ref={ref}
              emptyGrid={emptyGrid}
              setEmptyGrid={setEmptyGrid}
            />
          ))}
        </div>
      ))}
    </div>
  );
};

const ModePanel = ({ mode }) => (
  <div
    className="pb3"
    style={{
      background: "#F9F9F9"
    }}
    aria-live="polite"
  >
    {mode ? (
      <span>
        You are in{" "}
        <strong>{mode === MODE.brush ? "painting" : "erasing"}</strong> mode.
      </span>
    ) : null}
  </div>
);

export default App;
