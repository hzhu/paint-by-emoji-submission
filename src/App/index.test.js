import React from "react";
import {
  wait,
  render,
  cleanup,
  fireEvent,
  prettyDOM
} from "react-testing-library";
import App from "./index";
import {
  MAX_WIDTH,
  MAX_HEIGHT,
  MIN_WIDTH,
  MIN_HEIGHT,
  DEFAULT_WIDTH,
  DEFAULT_HEIGHT
} from "../constants";
import "jest-dom/extend-expect";

afterEach(cleanup);

test("renders the application correctly (snapshot test)", () => {
  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
});

test("user should be able to adjust size of the grid (e.g. from 10x8 to 3x3)", () => {
  // Given
  const NEW_WIDTH = 3;
  const NEW_HEIGHT = 3;
  const { getByLabelText, getByTestId } = render(<App />);
  const grid = getByTestId("grid");
  const widthInput = getByLabelText("Width");
  const heightInput = getByLabelText("Height");

  expect(grid.children.length).toBe(DEFAULT_HEIGHT); // grid starts with default height
  for (let i = 0; i < grid.children.length; i++) {
    const row = grid.children[i];
    expect(row.children.length).toBe(DEFAULT_WIDTH); // grid starts with default width
  }

  // When
  fireEvent.change(widthInput, { target: { value: NEW_WIDTH } });
  fireEvent.change(heightInput, { target: { value: NEW_HEIGHT } });

  // Then
  expect(grid.children.length).toBe(NEW_HEIGHT);
  for (let i = 0; i < grid.children.length; i++) {
    const row = grid.children[i];
    expect(row.children.length).toBe(NEW_WIDTH);
  }
});

test("user should see (or hear) text indicating current mode when the paint/brush button is selected", () => {
  // Given
  const PAINT_MODE = "You are in painting mode.";
  const ERASING_MODE = "You are in erasing mode.";
  const { getByAltText, queryAllByText } = render(<App />);
  const brushButton = getByAltText("brush");
  const eraseButton = getByAltText("eraser");
  var [modeElement] = queryAllByText(
    (_, { textContent }) => textContent === PAINT_MODE
  );
  expect(modeElement).toBe(undefined);

  // When
  fireEvent.click(brushButton);

  // Then
  var [modeElement] = queryAllByText(
    (_, { textContent }) => textContent === PAINT_MODE
  );
  expect(modeElement.textContent).toBe(PAINT_MODE);
  expect(modeElement).toHaveAttribute("aria-live", "polite");

  // And when
  fireEvent.click(eraseButton);

  // And then
  var [modeElement] = queryAllByText(
    (_, { textContent }) => textContent === ERASING_MODE
  );
  expect(modeElement.textContent).toBe(ERASING_MODE);
  expect(modeElement).toHaveAttribute("aria-live", "polite");
});

test("user should see (or hear) error message when changing grid size to be out of bounds", () => {
  // Given
  const { getByLabelText, queryAllByTestId } = render(<App />);
  const widthInput = getByLabelText("Width");
  const heightInput = getByLabelText("Height");
  const getBoundsError = () => queryAllByTestId("bounds-error");
  expect(getBoundsError()).toEqual([]);

  // When
  fireEvent.change(heightInput, { target: { value: 30 } });

  // Then
  var [heightError] = getBoundsError();
  expect(heightError.textContent).toBe(
    `⚠️ The maximum height is ${MAX_HEIGHT}`
  );

  // And when
  fireEvent.change(heightInput, { target: { value: -5 } });
  fireEvent.change(widthInput, { target: { value: 99 } });

  // And then
  var [widthError, heightError] = getBoundsError();
  expect(widthError.textContent).toBe(`⚠️ The maximum width is ${MAX_WIDTH}`);
  expect(heightError.textContent).toBe(
    `⚠️ The minimum height is ${MIN_HEIGHT}`
  );
  expect(widthError.parentElement).toHaveAttribute("aria-live", "polite");
  expect(heightError.parentElement).toHaveAttribute("aria-live", "polite");

  // And when
  fireEvent.change(widthInput, { target: { value: -5 } });

  // And then
  var [widthError] = getBoundsError();
  expect(widthError.textContent).toBe(`⚠️ The minimum width is ${MIN_WIDTH}`);
});

test("user should be able to paint (and erase) emojis on the grid", () => {
  // Given (a blank grid)
  const { getByTestId, getByAltText } = render(<App />);
  const grid = getByTestId("grid");
  const brushButton = getByAltText("brush");
  const eraseButton = getByAltText("eraser");
  const firstRowCoords = [[0, 0], [0, 1], [0, 2]];
  const getCell = ([m, n]) => getByTestId(`cell-${m}-${n}`);

  firstRowCoords.forEach(coord => expect(getCell(coord).textContent).toBe(""));

  // When (user paints in the first three cells of the first row)
  fireEvent.click(brushButton);
  fireEvent.mouseDown(grid);
  firstRowCoords.forEach(coord => fireEvent.mouseEnter(getCell(coord)));
  fireEvent.mouseUp(grid);

  // Then (the first row is painted with an emoji)
  firstRowCoords.forEach(coord =>
    expect(getCell(coord).textContent).toBe("😀")
  );

  // And when (user erases the first three cells of the first row)
  fireEvent.click(eraseButton);
  fireEvent.mouseDown(grid);
  firstRowCoords.forEach(coord => fireEvent.mouseEnter(getCell(coord)));
  fireEvent.mouseUp(grid);

  // And then (the first row is blank)
  firstRowCoords.forEach(coord => expect(getCell(coord).textContent).toBe(""));
});

test("user should be able to paint the canvas & copy emoji shortcodes to clipboard", async done => {
  // Given (user paints first three cells of the first row)
  const { getByText, getByTestId, getByAltText } = render(<App />);
  const grid = getByTestId("grid");
  const brushButton = getByAltText("brush");
  const firstRowCoords = [[0, 0], [0, 1], [0, 2]];
  const getCell = ([m, n]) => getByTestId(`cell-${m}-${n}`);
  firstRowCoords.forEach(coord => expect(getCell(coord).textContent).toBe(""));
  fireEvent.click(brushButton);
  fireEvent.mouseDown(grid);
  firstRowCoords.forEach(coord => fireEvent.mouseEnter(getCell(coord)));
  fireEvent.mouseUp(grid);
  firstRowCoords.forEach(coord =>
    expect(getCell(coord).textContent).toBe("😀")
  );
  console.log = jest.fn();
  const writeText = jest.fn(() => Promise.resolve());
  window.navigator.clipboard = { writeText };
  expect(writeText).toHaveBeenCalledTimes(0); // nothing should be in the clipboard
  expect(console.log).toHaveBeenCalledTimes(0);

  // When (user selects the copy button)
  const copyButton = getByText("Copy to clipboard");
  fireEvent.click(copyButton);

  // Then (emoji shortcodes are copied to their clipboard)
  const AFTER_CLIPBOARD = `:grinning::grinning::grinning::blank::blank::blank::blank::blank::blank::blank:
:blank::blank::blank::blank::blank::blank::blank::blank::blank::blank:
:blank::blank::blank::blank::blank::blank::blank::blank::blank::blank:
:blank::blank::blank::blank::blank::blank::blank::blank::blank::blank:
:blank::blank::blank::blank::blank::blank::blank::blank::blank::blank:
:blank::blank::blank::blank::blank::blank::blank::blank::blank::blank:
:blank::blank::blank::blank::blank::blank::blank::blank::blank::blank:
:blank::blank::blank::blank::blank::blank::blank::blank::blank::blank:`;

  expect(writeText).toHaveBeenCalledWith(AFTER_CLIPBOARD);
  expect(writeText).toHaveBeenCalledTimes(1);
  await wait(() => {
    expect(console.log).toHaveBeenCalledWith("Copied to clipboard");
    expect(console.log).toHaveBeenCalledTimes(1);
    done();
  });
});

test(`user should be able to clear the grid by selecting the "clear" button`, () => {
  // Given (user paints first three cells of the first row)
  const { getByText, getByTestId, getByAltText } = render(<App />);
  const grid = getByTestId("grid");
  const brushButton = getByAltText("brush");
  const firstRowCoords = [[0, 0], [0, 1], [0, 2]];
  const getCell = ([m, n]) => getByTestId(`cell-${m}-${n}`);
  firstRowCoords.forEach(coord => expect(getCell(coord).textContent).toBe(""));
  fireEvent.click(brushButton);
  fireEvent.mouseDown(grid);
  firstRowCoords.forEach(coord => fireEvent.mouseEnter(getCell(coord)));
  fireEvent.mouseUp(grid);
  firstRowCoords.forEach(coord =>
    expect(getCell(coord).textContent).toBe("😀")
  );

  // When (user selects the "Clear" button)
  const clearButton = getByText("Clear");
  fireEvent.click(clearButton);

  // Then (the painted cells are emptied)
  firstRowCoords.forEach(coord => expect(getCell(coord).textContent).toBe(""));
});

test('user is switched to "painting" mode when the erase button is selected & user clicks "clear"', () => {
  // Given (user paints first three cells of first row & selects erase mode)
  const PAINT_MODE = "You are in painting mode.";
  const ERASING_MODE = "You are in erasing mode.";
  const { getByText, getByTestId, getByAltText, queryAllByText } = render(
    <App />
  );
  const grid = getByTestId("grid");
  const brushButton = getByAltText("brush");
  const eraseButton = getByAltText("eraser");
  const firstRowCoords = [[0, 0], [0, 1], [0, 2]];
  const getCell = ([m, n]) => getByTestId(`cell-${m}-${n}`);
  firstRowCoords.forEach(coord => expect(getCell(coord).textContent).toBe(""));
  fireEvent.click(brushButton);
  var [modeElement] = queryAllByText(
    (_, { textContent }) => textContent === PAINT_MODE
  );
  expect(modeElement.textContent).toBe(PAINT_MODE);
  fireEvent.mouseDown(grid);
  firstRowCoords.forEach(coord => fireEvent.mouseEnter(getCell(coord)));
  fireEvent.mouseUp(grid);
  firstRowCoords.forEach(coord =>
    expect(getCell(coord).textContent).toBe("😀")
  );
  fireEvent.click(eraseButton);
  var [modeElement] = queryAllByText(
    (_, { textContent }) => textContent === ERASING_MODE
  );
  expect(modeElement.textContent).toBe(ERASING_MODE);

  // When (user selects the "Clear" button)
  const clearButton = getByText("Clear");
  fireEvent.click(clearButton);

  // Then (painted cells are emptied, user returned to painting mode)
  firstRowCoords.forEach(coord => expect(getCell(coord).textContent).toBe(""));
  var [modeElement] = queryAllByText(
    (_, { textContent }) => textContent === PAINT_MODE
  );
  expect(modeElement.textContent).toBe(PAINT_MODE);
});
