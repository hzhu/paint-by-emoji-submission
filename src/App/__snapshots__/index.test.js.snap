// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the application correctly (snapshot test) 1`] = `
<div>
  <div
    class="app flex items-center flex-column justify-center h-100"
  >
    <div
      class="bg-white shadow-x-1 tc br3 ba b--light-silver overflow-hidden"
    >
      <div
        class="bg-light-gray flex tb-padding"
      >
        <div
          class="flex-auto relative"
        >
          <div
            style="display: inline;"
          >
            <span
              class="visually-hidden"
              id="emoji_btn_label"
            >
              Choose an emoji
            </span>
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="emoji_btn_label exp_button"
              class="br2 fl pointer b--light-silver ep-btn"
              id="exp_button"
              tabindex="0"
            >
              ðŸ˜€
            </button>
          </div>
          <button
            class="ba br2 b--light-silver pointer dib fl overflow-hidden ep-ctrl "
          >
            <img
              alt="brush"
              class="h-100"
              src="brush.png"
            />
          </button>
          <button
            class="ba br2 b--light-silver pointer dib fl overflow-hidden ep-ctrl "
          >
            <img
              alt="eraser"
              class="h-100"
              src="eraser.png"
            />
          </button>
        </div>
        <div>
          <label
            class="emoji-paint__dimension b ml3"
          >
            Width
            <input
              class="ml2 f6 ba b--light-silver br2 tc tb-ctrl-size"
              max="30"
              min="1"
              type="number"
              value="10"
            />
          </label>
          <label
            class="emoji-paint__dimension b ml3"
          >
            Height
            <input
              class="ml2 f6 ba b--light-silver br2 tc tb-ctrl-size"
              max="20"
              min="1"
              type="number"
              value="8"
            />
          </label>
          <div
            class="dark-red pt2 tl"
          >
            <div
              aria-live="polite"
            />
            <div
              aria-live="polite"
            />
          </div>
        </div>
      </div>
      <div
        aria-live="polite"
        class="pb3"
        style="background: rgb(249, 249, 249);"
      />
      <div
        class="ma3 br3 dib b--gray not-allowed"
        data-testid="grid"
      >
        <div
          class="flex"
          data-testid="row-0"
        >
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-0"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-1"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-2"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-3"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-4"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-5"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-6"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-7"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-8"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-0-9"
            role="img"
          />
        </div>
        <div
          class="flex"
          data-testid="row-1"
        >
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-0"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-1"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-2"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-3"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-4"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-5"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-6"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-7"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-8"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-1-9"
            role="img"
          />
        </div>
        <div
          class="flex"
          data-testid="row-2"
        >
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-0"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-1"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-2"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-3"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-4"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-5"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-6"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-7"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-8"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-2-9"
            role="img"
          />
        </div>
        <div
          class="flex"
          data-testid="row-3"
        >
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-0"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-1"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-2"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-3"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-4"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-5"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-6"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-7"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-8"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-3-9"
            role="img"
          />
        </div>
        <div
          class="flex"
          data-testid="row-4"
        >
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-0"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-1"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-2"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-3"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-4"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-5"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-6"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-7"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-8"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-4-9"
            role="img"
          />
        </div>
        <div
          class="flex"
          data-testid="row-5"
        >
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-0"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-1"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-2"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-3"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-4"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-5"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-6"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-7"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-8"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-5-9"
            role="img"
          />
        </div>
        <div
          class="flex"
          data-testid="row-6"
        >
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-0"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-1"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-2"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-3"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-4"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-5"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-6"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-7"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-8"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-6-9"
            role="img"
          />
        </div>
        <div
          class="flex"
          data-testid="row-7"
        >
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-0"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-1"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-2"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-3"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-4"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-5"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-6"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-7"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-8"
            role="img"
          />
          <span
            aria-label="blank emoji"
            class="pa1 usn cell"
            data-testid="cell-7-9"
            role="img"
          />
        </div>
      </div>
      <div
        class="flex justify-end mh3"
      >
        <button
          class="ph3 br2 fw5 pointer b--light-silver"
        >
          Clear
        </button>
        <button
          class="f6 fw6 ml2 white ph3 pv3 bn br2 pointer bg-light-green"
        >
          Copy to clipboard
        </button>
      </div>
      <div
        class="pa3"
      >
        <span>
          Learn how to add a 
          <em>
            :blank:
          </em>
           emoji to your Slack workspace
           
          <a
            class="link bg-animate hover-bg-lightest-blue"
            href="https://get.slack.help/hc/en-us/articles/206870177-Add-custom-emoji"
          >
            here!
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;
